src/hotspot/share/gc/shared/softRefGenPolicy.cpp

```diff
@@ -1,40 +0,0 @@
-/*
- * Copyright (c) 2001, 2023, Oracle and/or its affiliates. All rights reserved.
- * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
- *
- * This code is free software; you can redistribute it and/or modify it
- * under the terms of the GNU General Public License version 2 only, as
- * published by the Free Software Foundation.
- *
- * This code is distributed in the hope that it will be useful, but WITHOUT
- * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
- * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
- * version 2 for more details (a copy is included in the LICENSE file that
- * accompanied this code).
- *
- * You should have received a copy of the GNU General Public License version
- * 2 along with this work; if not, write to the Free Software Foundation,
- * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
- *
- * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
- * or visit www.oracle.com if you need additional information or have any
- * questions.
- *
- */
-
-#include "precompiled.hpp"
-#include "gc/shared/adaptiveSizePolicy.hpp"
-#include "gc/shared/genCollectedHeap.hpp"
-#include "gc/shared/softRefGenPolicy.hpp"
-
-void SoftRefGenPolicy::cleared_all_soft_refs() {
-  // If near gc overhear limit, continue to clear SoftRefs.  SoftRefs may
-  // have been cleared in the last collection but if the gc overhear
-  // limit continues to be near, SoftRefs should still be cleared.
-  AdaptiveSizePolicy* size_policy = GenCollectedHeap::heap()->size_policy();
-  if (size_policy != nullptr) {
-    set_should_clear_all_soft_refs(size_policy->gc_overhead_limit_near());
```

Comment:
Why is this call removed? I'd expect `set_should_clear_all_soft_refs(false);` here.

Reply:
The default value of `_should_clear_all_soft_refs` is `false` and I don't see other places to set it to `true` in Serial GC. And from the comment, I think it is a leftover of CMS.

>  // If near gc overhear limit, continue to clear SoftRefs.  SoftRefs may
>  // have been cleared in the last collection but if the gc overhear
>  // limit continues to be near, SoftRefs should still be cleared.

Reply:
The `WB_FullGC` is just test code. I have several problems. If the test code wants the GC to clear all soft references (`set_should_clear_all_soft_refs(true)` in `WB_FullGC`), why do we set it to `false` again? If we have a reason to set it to `false` again, why do other GCs not set it (Why is `SoftRefGenPolicy` only used at Serial GC?).

> // If near gc overhear limit, continue to clear SoftRefs. SoftRefs may
> // have been cleared in the last collection but if the gc overhear
> // limit continues to be near, SoftRefs should still be cleared.

From the comment, the intent of the method `SoftRefGenPolicy::cleared_all_soft_refs` is letting the next GC re-clear the soft reference if **near gc overhear limit**. But the Serial GC has no gc overhear limit check now. So I still think it is just a leftover of CMS.

Reply:
> why do we set it to false again

`WB_FullGC` should restore the global env. Otherwise, the following non-whitebox-full-gc cycles can unnecessarily clear soft-refs.

> If we have a reason to set it to false again, why do other GCs not set it

G1 does that inside `WB_FullGC`. ZGC doesn't rely on `SoftRefPolicy::should_clear_all_soft_refs`. Parallel doesn't clear the flag, which I believe is a bug.

This inconsistency is probably accidental; keeping the global env intact (essentially what ZGC does) seems the cleanest approach IMO. I wonder if other collectors can also adopt it; in another PR(s) ofc.

Reply:
> > why do we set it to false again
> 
> `WB_FullGC` should restore the global env. Otherwise, the following non-whitebox-full-gc cycles can unnecessarily clear soft-refs.
>
> > If we have a reason to set it to false again, why do other GCs not set it
> 
> G1 does that inside `WB_FullGC`. ZGC doesn't rely on `SoftRefPolicy::should_clear_all_soft_refs`. Parallel doesn't clear the flag, which I believe is a bug.
> 
> This inconsistency is probably accidental; keeping the global env intact (essentially what ZGC does) seems the cleanest approach IMO. I wonder if other collectors can also adopt it; in another PR(s) ofc.

Thanks for your answers. I propose the following changes.

In this patch:
- Serial GC should use `SoftRefPolicy` instead of `SoftRefGenPolicy`, which is done in this patch now.
- Serial GC should restore the global env inside `WB_FullGC`, just like the G1. Because both of them don't implement overheap limit check now.
- Don't remove the class `SoftRefGenPolicy` because it will be used by Parallel GC in the future. 

In another patch:
- Rename `SoftRefGenPolicy` in order to avoid misunderstanding, such as `GCOverhearLimitSoftRefPolicy`.
- Use `SoftRefGenPolicy` (may be renamed at the first step) in Parallel GC.

Reply:
>  I will file new tickets to follow up.

Filed https://bugs.openjdk.org/browse/JDK-8309616 and https://bugs.openjdk.org/browse/JDK-8309619 to follow up.